src/main/java/com/hivemq/mqtt/handler/InboundInterceptorHandler.java

```diff
@@ -0,0 +1,79 @@
+import com.hivemq.mqtt.message.disconnect.DISCONNECT;
+import com.hivemq.mqtt.message.puback.PUBACK;
+import com.hivemq.mqtt.message.pubcomp.PUBCOMP;
+import com.hivemq.mqtt.message.pubrec.PUBREC;
+import com.hivemq.mqtt.message.pubrel.PUBREL;
+import com.hivemq.mqtt.message.unsubscribe.UNSUBSCRIBE;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * @author Daniel Krüger
+ */
+@Singleton
+@ChannelHandler.Sharable
+public class InboundInterceptorHandler extends ChannelInboundHandlerAdapter {
+
+    private final @NotNull ConnectInboundInterceptorHandler connectInboundInterceptorHandler;
+    private final @NotNull UnsubscribeInboundInterceptorHandler unsubscribeInboundInterceptorHandler;
+
+    private final @NotNull PingInterceptorHandler pingInterceptorHandler;
+    private final @NotNull PubackInterceptorHandler pubackInterceptorHandler;
+    private final @NotNull PubrecInterceptorHandler pubrecInterceptorHandler;
+    private final @NotNull PubcompInterceptorHandler pubcompInterceptorHandler;
+    private final @NotNull PubrelInterceptorHandler pubrelInterceptorHandler;
+
+    private final @NotNull DisconnectInterceptorHandler disconnectInterceptorHandler;
+
+    @Inject
+    public InboundInterceptorHandler(final @NotNull ConnectInboundInterceptorHandler connectInboundInterceptorHandler,
+                                     final @NotNull UnsubscribeInboundInterceptorHandler unsubscribeInboundInterceptorHandler,
+                                     final @NotNull PingInterceptorHandler pingInterceptorHandler,
+                                     final @NotNull PubackInterceptorHandler pubackInterceptorHandler,
+                                     final @NotNull PubrecInterceptorHandler pubrecInterceptorHandler,
+                                     final @NotNull PubcompInterceptorHandler pubcompInterceptorHandler,
+                                     final @NotNull PubrelInterceptorHandler pubrelInterceptorHandler,
+                                     final @NotNull DisconnectInterceptorHandler disconnectInterceptorHandler) {
+        this.connectInboundInterceptorHandler = connectInboundInterceptorHandler;
+        this.unsubscribeInboundInterceptorHandler = unsubscribeInboundInterceptorHandler;
+        this.pingInterceptorHandler = pingInterceptorHandler;
+        this.pubackInterceptorHandler = pubackInterceptorHandler;
+        this.pubrecInterceptorHandler = pubrecInterceptorHandler;
+        this.pubcompInterceptorHandler = pubcompInterceptorHandler;
+        this.pubrelInterceptorHandler = pubrelInterceptorHandler;
+        this.disconnectInterceptorHandler = disconnectInterceptorHandler;
+    }
+
+    @Override
+    public void channelRead(final @NotNull ChannelHandlerContext ctx, final @NotNull Object msg) {
+        if (msg instanceof CONNECT) {
```

Comment:
Imho it would be really great if we could avoid the multiple ifs and the instanceof checks.
For that we should only check for `instanceof Message` and utilize the `MessageType` to index into an array. We would then also need a common interface for the interceptor handlers that takes `Message` as an argument and casts internally.

Reply:
if the array contains null, then directly call `ctx.fireChannelRead(msg);`

Reply:
well this has a constant complexity against a linear complexity - thats true. The Handling itself is more complex imho.